design/appfile-design.md

```diff
@@ -0,0 +1,337 @@
+  name: webservice
+spec:
+  definitionRef:
+    name: deployments.apps
+  extension:
+    template: |
+      parameter: #webservice
+      #webservice: {
+        // +vela:cli:enabled=true
+        // +vela:cli:usage=specify commands to run in container
+        // +vela:cli:short=c
+        cmd: [...string]
+      
+        env: [...string]
+      
+        files: [...string]
+      }
+      
+      output: {
+        apiVersion: "apps/v1"
+        kind: "Deployment"
+        metadata:
+          name: context.name
+        spec: {
+          selector: {
+            matchLabels:
+              app: context.name
+          }
+          template: {
+            metadata:
+              labels:
+                app: context.name
+            spec: {
+              containers: [{
+                name:  context.name
+                image: context.image
+                command: parameter.cmd
+              }]
+            }
+          }
+        }
+      }
+```
+
+Apply the file to APIServer, and the fields will be extended into Appfile. Note that there are some conventions that differs Workloads and Traits, and around CLI flags. We will cover that more detailedly below.
+
+## Goals
+
+The Appfile design has the following goals:
+
+1. Provide a user friendly, `docker-compose` style config format to developers.
+2. Configuration fields can be extended by registering more capabilities into OAM runtime.
+
+In the following, we will discuss technical details of the proposed design.
+
+## Proposal
+
+### Registration via Definition/Capability
+
+Vela allows platform builders to extend Appfile config fields by registering them via [capabilities system of OAM](https://github.com/oam-dev/kubevela/blob/master/DESIGN.md#capability-register-and-discovery).
+
+The entire template should be put under `spec.extension.template` as raw string:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition | TraitDefinition
+...
+spec:
+  extension:
+    template: |
+      parameter: #webservice
+      ...
+```
+
+By running `vela system update` or other similar commands, vela cli will read all definitions from APIServer and sink necessary information locally including templates. The templates will be further used to render final deploy manifests.
+
+### Templating
+
+Vela allows platform builders to write bespoke templates to extend Appfile configs. 
+
+#### Exposing Parameters
+
+A template starts with `parameter` and its definition:
+
+```yaml
+parameter: #webservice
+#webservice: {
+  // +vela:cli:enabled=true
+  // +vela:cli:usage=specify commands to run in container
+  // +vela:cli:short=c
+  cmd: [...string]
+}
+```
+
+Here is the takeout:
+* The `parameter` defines the user input fields and is used to render final output with user input values. These fields will be exposed to users in Appfile.
+* The definition `#webservice` is used to tell the name of the template. This name is used to correlate workload and trait types to fields in Appfile.
```

Comment:
Please clarify in the appfile side about how the developer pick workload type.

Reply:
This is clarified in the `type` field below.